# Ratac package vignette

Let's define some preliminary variables

```{r}
    `%>%` <- magrittr::`%>%`
    `%<>%` <- magrittr::`%<>%`
    order.tissues <- c('Germline', 'Neurons', 'Muscle', 'Hypod.', 'Intest.')
    col.vec <- c( '#1232D9', '#3B9B46', '#D99B12', '#9e9e9e', '#D912D4', '#9E7FCC', '#B0E797', '#D1B3B3', '#991919', '#23A4A3', '#000000', '#dbdbdb')
```

## Import reads bam files

We can read all the .bam files within one folder using the `importBam()` function.

```{r}
    # Find .bam files
    bam.folder <- '~/20190118_ATAC-seq_PE-mapping/_bam-files'
    bam.files <- list.files(bam.folder, pattern = 'combined.bam$', full.names = T)[c(1, 6, 5, 2, 3)]
    # Read through .bam files
    bam.list <- importBam(
        bam.files, 
        remove.duplicates = T, 
        paired.reads = T, 
        filter.lengths = c(50, 1000), 
        quality.cutoff = 10
    ) %>% stats::setNames(basename(bam.files) %>% gsub('Gonad', 'Germline', .) %>% gsub('_YA_ATAC_combined.bam', '', .))
    # We can create a 'pseudo-Soma' sample containing only reads from somatic tissues
    bam.list[['pseudo-Soma']] <- unlist(GRangesList(lapply(order.tissues[2:5], function(TISSUE) {
        granges <- bam.list[[TISSUE]]
        return(granges[sample(x = 1:length(granges), size = 10000000, replace = F)])
    } )))
    # Save data
    saveRDS(bam.list, file = 'data/bam.list.rds')
    save(bam.list, file = 'data/bam.list.RData')
    load('bam.list')
    #bam.list <- readRDS('data/bam.list.rds')
```

## Check distribution of fragment sizes over a specific set of segments, for each file. 

Let's have a look at the distribution of fragment sizes in each ATAC-seq library, over different classes of REs.

```{r}
    # Get tissue-specific REs
    REs <- rtracklayer::import('~/20190412_classification-REs/_results-files/tissue.specific.ATAC_RE.max.tissues.FC-3_TPM-20_final-annotations.gff3')
    GenomicRanges::mcols(REs) %<>% as.data.frame %>% tibble::as_tibble() %>% dplyr::select(c("type", "ID", "RE_idx", "Name", "associated.gene.WB", "domain",  "which.tissue"))
    proms.list <- list(
        "Germline_REs" = REs[REs$type %in% c('fwd-promoter', 'rev-promoter', 'bidirect-promoter') & REs$which.tissue == 'Germline'],
        "Neurons_REs" = REs[REs$type %in% c('fwd-promoter', 'rev-promoter', 'bidirect-promoter') & REs$which.tissue == 'Neurons'],
        "Muscle_REs" = REs[REs$type %in% c('fwd-promoter', 'rev-promoter', 'bidirect-promoter') & REs$which.tissue == 'Muscle'],
        "Hypod._REs" = REs[REs$type %in% c('fwd-promoter', 'rev-promoter', 'bidirect-promoter') & REs$which.tissue == 'Hypod'],
        "Intest._REs" = REs[REs$type %in% c('fwd-promoter', 'rev-promoter', 'bidirect-promoter') & REs$which.tissue == 'Intest'],
        "Soma_REs" = REs[REs$type %in% c('fwd-promoter', 'rev-promoter', 'bidirect-promoter') & REs$which.tissue == 'Soma'],
        "Ubiq._REs" = REs[REs$type %in% c('fwd-promoter', 'rev-promoter', 'bidirect-promoter') & REs$which.tissue == 'Ubiq']
    )
    # For each bam.file, plot its distribution of fragment sizes over the tissue-specific REs or ubiquitous REs
    sizes <- parallel::mclapply(c('Germline', 'pseudo-Soma'), function(TISSUE) {
        getFragmentsDistribution(bam.list[[TISSUE]], proms.list)
    }, mc.cores = 2) %>% setNames(c('Germline', 'pseudo-Soma'))
    # Plot results
    distribution_fragments_sizes.list <- plotFragmentsDistribution(sizes)
    p <- cowplot::plot_grid(plotlist = distribution_fragments_sizes.list, ncol = 2)
    ggplot2::ggsave('examples/pdf/fragments_size_distribution.pdf', height = 10, width = 20)
````

We can already notice that there is a lack of second/third/etc. nucleosome signature in somatic-specific ATAC-seq data over somatic REs. 
However, a more in-depth view of nucleosome binding would be desirable, to better understand the molecular context of promoter organization.

## Vmats

Particularly, *where* fragments are mapping is an information as important as *how long* they are. 
For this reason, we created a function to generate V-plots to show distribution of ATAC fragments 
of variable sizes around a stack of regions of interest.

```{r}
    breaks <- seq(0, 25, length.out = 61)
    Vmat <- bam.list[['Neurons']] %>% 
        computeVmat(proms.list[['Neurons_REs']]) %>%
        normalizeVmat(normFUN = 'pctsum', roll = 3)
    plotVmat(
        Vmat,
        breaks = breaks, 
        main = 'Neurons-spe. ATAC fragments over Neurons-spe. proms', 
        pdf = 'examples/pdf/Vmat_Neurons-over-neurons-REs.pdf'
    )
    # Or do all at once:
    neurons_atac_neurons_REs <- plotVmat(
        bam.list[['Neurons']], 
        proms.list[['Neurons_REs']], 
        breaks = breaks, 
        main = 'Neurons-spe. ATAC fragments over Neurons-spe. proms', 
    )
    gl_atac_gl_REs <- plotVmat(
        bam.list[['Germline']], 
        proms.list[['Germline_REs']], 
        breaks = breaks, 
        main = 'Germline-spe. ATAC fragments over Germline-spe. proms', 
    )
    plots <- cowplot::plot_grid(gl_atac_gl_REs, neurons_atac_neurons_REs)
    ggplot2::ggsave('examples/pdf/GL_neurons_Vmat-comparison.pdf', height = 6, width = 12, plots)
```

## Normalized Vmats

Ideally, one would take into account the background ATAC-seq fragments to better estimate the molecular context in each condition.
Samples might have an background fragment size higher than others and this could affect the appearance of un-normalized Vmats.
To estimate background, simply use the `estimate.background = T` argument. 

```{r}
    neurons_atac_neurons_REs_NORM <- plotVmat(
        bam.list[['Neurons']], 
        proms.list[['Neurons_REs']], 
        estimate.background = TRUE,
        breaks = breaks, 
        main = 'Neurons-spe. ATAC fragments over Neurons-spe. proms (Norm.)'
    )
    gl_atac_gl_REs_NORM <- plotVmat(
        bam.list[['Germline']], 
        proms.list[['Germline_REs']], 
        estimate.background = TRUE,
        breaks = breaks, 
        main = 'Germline-spe. ATAC fragments over Germline-spe. proms (Norm.)'
    )
    plots <- cowplot::plot_grid(gl_atac_gl_REs_NORM, neurons_atac_neurons_REs_NORM)
    ggplot2::ggsave('examples/pdf/GL_neurons_normalized-Vmat-comparison.pdf', height = 6, width = 12, plots)
    # This is equivalent to:
    Vmat2 <- computeVmat(
        bam.list[['Neurons']], 
        shuffleGRanges(proms.list[['Neurons_REs']])
    )
    Vmat <- bam.list[['Neurons']] %>% 
        computeVmat(proms.list[['Neurons_REs']]) %>%
        normalizeVmat(Vmat2 = Vmat2, normalize = T, normFUN = 'pctsum', roll = 3)
    plotVmat(
        Vmat,
        breaks = breaks, 
        main = 'Neurons-spe. ATAC fragments over Neurons-spe. proms', 
        pdf = 'examples/pdf/Vmat-normalized_Neurons-over-neurons-REs.pdf'
    )
```

## Nucleosome enrichment estimation

```{r}
    pseudosoma_atac_somatic_REs_NORM <- plotVmat(
        bam.list[['pseudo-Soma']], 
        proms.list[2:5] %>% GenomicRanges::GRangesList() %>% unlist(), 
        estimate.background = TRUE,
        breaks = breaks, 
        main = 'pseudoSoma ATAC fragments over Somatic proms (Norm.)'
    )
    pseudosoma_atac_ubiq_REs_NORM <- plotVmat(
        bam.list[['pseudo-Soma']], 
        proms.list[['Ubiq._REs']], 
        estimate.background = TRUE,
        breaks = breaks, 
        main = 'pseudoSoma ATAC fragments over Ubiquitous proms (Norm.)'
    )
    germline_atac_germline_REs_NORM <- plotVmat(
        bam.list[['Germline']], 
        proms.list[['Germline_REs']], 
        estimate.background = TRUE,
        breaks = breaks, 
        main = 'Germline ATAC fragments over Germline proms (Norm.)'
    )
    germline_atac_ubiq_REs_NORM <- plotVmat(
        bam.list[['Germline']], 
        proms.list[['Ubiq._REs']], 
        estimate.background = TRUE,
        breaks = breaks, 
        main = 'Germline ATAC fragments over Ubiquitous proms (Norm.)'
    )
    plots <- cowplot::plot_grid(
        pseudosoma_atac_somatic_REs_NORM, 
        pseudosoma_atac_ubiq_REs_NORM, 
        germline_atac_germline_REs_NORM, 
        germline_atac_ubiq_REs_NORM,
        ncol = 2, 
        nrow = 2
    )
    ggplot2::ggsave('examples/pdf/GL-pseudosoma_normalized-Vmat-comparison.pdf', height = 12, width = 12, plots)
```

From this figure, it is striking that Germline REs are characterized by prominent flanking nucleosomes in Germline, 
while somatic REs in somatic tissues are not. Ubiquitous REs, on the other hand, 
are flanked by nucleosomes in all tissues. 

We can estimate nucleosome enrichment using the `nucleosomeEnrichment()` function.

```{r}
    Vmat_gl_over_gl <- nucleosomeEnrichment(
        bam.granges = bam.list[['Germline']], 
        granges = proms.list[['Germline_REs']], 
        estimate.background = TRUE)
    Vmat_gl_over_ubiq <- nucleosomeEnrichment(
        bam.list[['Germline']], 
        proms.list[['Ubiq._REs']], 
        estimate.background = TRUE
    )
    Vmat_psoma_over_psoma <- nucleosomeEnrichment(
        bam.list[['pseudo-Soma']], 
        proms.list[2:5] %>% GenomicRanges::GRangesList() %>% unlist(), 
        estimate.background = TRUE
    )
    Vmat_psoma_over_ubiq <- nucleosomeEnrichment(
        bam.list[['pseudo-Soma']], 
        proms.list[['Ubiq._REs']], 
        estimate.background = TRUE
    )
    Vmat_nuc.enrich <- list(gl_over_gl, gl_over_ubiq, psoma_over_psoma, psoma_over_ubiq)
    barplot(sapply(nuc.enrich, function(Vmat) attributes(Vmat)$fisher$estimate), col = 'white')
    abline(h = 1, lty = 2, col = 'red')
```

## Real case use: Check nucleosome enrichment over tissue-specific enhancers

```{r}
    # Get tissue-specific REs
    REs <- rtracklayer::import('~/20190412_classification-REs/_results-files/tissue.specific.ATAC_RE.max.tissues.FC-3_TPM-20_final-annotations.gff3')
    GenomicRanges::mcols(REs) %<>% as.data.frame %>% tibble::as_tibble() %>% dplyr::select(c("type", "ID", "RE_idx", "Name", "associated.gene.WB", "domain",  "which.tissue"))
    enhs.list <- list(
        "Germline_REs" = REs[REs$type %in% c('putative_enhancer') & REs$which.tissue == 'Germline'],
        "Neurons_REs" = REs[REs$type %in% c('putative_enhancer') & REs$which.tissue == 'Neurons'],
        "Muscle_REs" = REs[REs$type %in% c('putative_enhancer') & REs$which.tissue == 'Muscle'],
        "Hypod._REs" = REs[REs$type %in% c('putative_enhancer') & REs$which.tissue == 'Hypod'],
        "Intest._REs" = REs[REs$type %in% c('putative_enhancer') & REs$which.tissue == 'Intest'],
        "Soma_REs" = REs[REs$type %in% c('putative_enhancer') & REs$which.tissue == 'Soma'],
        "Ubiq._REs" = REs[REs$type %in% c('putative_enhancer') & REs$which.tissue == 'Ubiq']
    )

```

## Bonus 1: metaplot function

It is often interesting to aggregate many different regions into one "meta-region" 
and plot average signal of ChIP-seq, MNase-seq, ATAC-seq, etc. over this meta-region.
This can be easily achieved here using the `plotMetaCoverage()` function.

```{r}
    # Get tissue-specific ATAC-seq data
    bw.folder <- '~/__ATAC_mapping/_bw-files'
    bw.files <- list.files(bw.folder, pattern = 'combined.bw$', full.names = T)[c(1, 5, 4, 2, 3)]
    bw.list <- importBigWig(bw.files) %>% 
        stats::setNames(basename(bw.files) %>% gsub('Gonad', 'Germline', .) %>% gsub('_YA_ATAC_combined.bw', '', .))
    # Get tissue-specific REs
    REs <- rtracklayer::import('~/20190412_classification-REs/_results-files/tissue.specific.ATAC_RE.max.tissues.FC-3_TPM-20_final-annotations.gff3')
    GenomicRanges::mcols(REs) %<>% as.data.frame %>% tibble::as_tibble() %>% dplyr::select(c("type", "ID", "RE_idx", "Name", "associated.gene.WB", "domain",  "which.tissue"))
    proms.list <- list(
        "Germline_REs" = REs[REs$type %in% c('fwd-promoter', 'rev-promoter', 'bidirect-promoter') & REs$which.tissue == 'Germline'],
        "Neurons_REs" = REs[REs$type %in% c('fwd-promoter', 'rev-promoter', 'bidirect-promoter') & REs$which.tissue == 'Neurons'],
        "Muscle_REs" = REs[REs$type %in% c('fwd-promoter', 'rev-promoter', 'bidirect-promoter') & REs$which.tissue == 'Muscle'],
        "Hypod._REs" = REs[REs$type %in% c('fwd-promoter', 'rev-promoter', 'bidirect-promoter') & REs$which.tissue == 'Hypod'],
        "Intest._REs" = REs[REs$type %in% c('fwd-promoter', 'rev-promoter', 'bidirect-promoter') & REs$which.tissue == 'Intest'],
        "Soma_REs" = REs[REs$type %in% c('fwd-promoter', 'rev-promoter', 'bidirect-promoter') & REs$which.tissue == 'Soma'],
        "Ubiq._REs" = REs[REs$type %in% c('fwd-promoter', 'rev-promoter', 'bidirect-promoter') & REs$which.tissue == 'Ubiq']
    ) %>% GenomicRanges::GRangesList()
    # Plot Germline over each Germline promoters
    file <- '~/__ATAC_mapping/_bw-files/Intest._YA_ATAC_combined.bw'
    p <- coverageMetaplot(file, name = 'Intestine', REs[REs$which.tissue == 'Intest'], colors = '#D912D4')
    # Plot Germline over class of promoters
    plots <- coverageMetaplot(bw.list[[1]], name = 'Germline', proms.list, colors = 'blue', ylim = c(0, 25))
    g <- cowplot::plot_grid(plotlist = plots, nrow = 1)
    ggplot2::ggsave('examples/pdf/ATAC-seq_germline.pdf', width = 15, height = 3, g)
    # Plot Germline over each class of promoters
    plots <- coverageMetaplot(bw.list, proms.list, colors = col.vec)
    g <- cowplot::plot_grid(plotlist = plots, nrow = 1)
    ggplot2::ggsave('examples/pdf/ATAC-seq-coverage_over_classes-of-proms.pdf', width = 15, height = 3, g)
    # Plot reverse
    plots <- coverageMetaplot(bw.list, proms.list, colors = col.vec, by.granges = F)
    g <- cowplot::plot_grid(plotlist = plots, nrow = 1)
    ggplot2::ggsave('examples/pdf/ATAC-seq-coverage_proms-coverage.pdf', width = 15, height = 3, g)
```


