<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>A function to import paired end bam files as GRanges — importPEBamFiles • VplotR</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="A function to import paired end bam files as GRanges — importPEBamFiles" />
<meta property="og:description" content="This function takes bam file paths and read them into GRanges 
objects.
Note: Can be quite lengthy for .bam files with 5+ millions fragments." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">VplotR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/VplotR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/js2264/VplotR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>A function to import paired end bam files as GRanges</h1>
    <small class="dont-index">Source: <a href='https://github.com/js2264/VplotR/blob/master/R/bamutils.R'><code>R/bamutils.R</code></a></small>
    <div class="hidden name"><code>importPEBamFiles.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function takes bam file paths and read them into GRanges 
objects.
Note: Can be quite lengthy for .bam files with 5+ millions fragments.</p>
    </div>

    <pre class="usage"><span class='fu'>importPEBamFiles</span>(
  <span class='no'>files</span>,
  <span class='kw'>genome</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>where</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>max_insert_size</span> <span class='kw'>=</span> <span class='fl'>1000</span>,
  <span class='kw'>shift_ATAC_fragments</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>cores</span> <span class='kw'>=</span> <span class='fl'>10</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>files</th>
      <td><p>character vector, each element of the vector is the path 
of an individual .bam file.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>character, genome ID (e.g. sacCer3, ce11, dm6, danRer10,
mm10 or hg38).</p></td>
    </tr>
    <tr>
      <th>where</th>
      <td><p>GRanges, only import the fragments mapping to the 
input GRanges (can fasten the import process a lot).</p></td>
    </tr>
    <tr>
      <th>max_insert_size</th>
      <td><p>Integer, filter out fragments larger 
than this size.</p></td>
    </tr>
    <tr>
      <th>shift_ATAC_fragments</th>
      <td><p>Boolean, if the fragments come 
from ATAC-seq, one might want to shift the extremities by +5 / -4 bp.</p></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>Integer, number of cores to use when indexing bam files</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Boolean</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A GRanges object containing fragments from the input .bam file.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>bamfile</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"ex1.bam"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"Rsamtools"</span>)
<span class='no'>fragments</span> <span class='kw'>&lt;-</span> <span class='fu'>importPEBamFiles</span>(
    <span class='no'>bamfile</span>,
    <span class='kw'>shift_ATAC_fragments</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
)</div><div class='output co'>#&gt; <span class='message'>&gt; Importing /home/travis/R/Library/Rsamtools/extdata/ex1.bam ...</span></div><div class='output co'>#&gt; <span class='message'>&gt; Filtering /home/travis/R/Library/Rsamtools/extdata/ex1.bam ...</span></div><div class='output co'>#&gt; <span class='message'>&gt; Shifting /home/travis/R/Library/Rsamtools/extdata/ex1.bam ...</span></div><div class='output co'>#&gt; <span class='message'>&gt; /home/travis/R/Library/Rsamtools/extdata/ex1.bam import completed.</span></div><div class='input'><span class='no'>fragments</span></div><div class='output co'>#&gt; GRanges object with 1572 ranges and 0 metadata columns:
#&gt;          seqnames    ranges strand
#&gt;             &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
#&gt;      [1]     seq1    41-215      +
#&gt;      [2]     seq1    54-255      +
#&gt;      [3]     seq1    56-258      +
#&gt;      [4]     seq1    65-255      +
#&gt;      [5]     seq1    65-265      +
#&gt;      ...      ...       ...    ...
#&gt;   [1568]     seq2 1326-1542      -
#&gt;   [1569]     seq2 1336-1544      -
#&gt;   [1570]     seq2 1358-1550      -
#&gt;   [1571]     seq2 1380-1557      -
#&gt;   [1572]     seq2 1353-1562      -
#&gt;   -------
#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://js2264.github.io/'>Jacques Serizay</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


