<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>VplotR advanced functions • VplotR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="VplotR advanced functions">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">VplotR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/VplotR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/advanced.html">In-depth use</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>VplotR advanced functions</h1>
                        <h4 class="author">Jacques Serizay</h4>
            
            <h4 class="date">2020-02-09</h4>
      
      
      <div class="hidden name"><code>advanced.Rmd</code></div>

    </div>

    
    
<div id="set-up-the-session" class="section level2">
<h2 class="hasAnchor">
<a href="#set-up-the-session" class="anchor"></a>0. Set up the session</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(tidyverse)</a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(VplotR)</a>
<a class="sourceLine" id="cb1-3" title="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(ggplot2)</a></code></pre></div>
</div>
<div id="import-reads-bam-files-and-genomic-loci-files" class="section level2">
<h2 class="hasAnchor">
<a href="#import-reads-bam-files-and-genomic-loci-files" class="anchor"></a>1. Import reads bam files and genomic loci files</h2>
<p>We can import .bam files of interest from a given folder using the <code>importBamFiles()</code> function.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="co"># Import bam file over annotated promoters, from a local folder. </span></a>
<a class="sourceLine" id="cb2-2" title="2"><span class="co"># </span><span class="al">WARNING</span><span class="co">: This takes quite a long time for libraries that are deeply sequenced</span></a>
<a class="sourceLine" id="cb2-3" title="3">bam_files &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="st">'~/path/to/_bam-files/'</span>)</a>
<a class="sourceLine" id="cb2-4" title="4">bam_list &lt;-<span class="st"> </span><span class="kw"><a href="../reference/importPEBamFiles.html">importPEBamFiles</a></span>(</a>
<a class="sourceLine" id="cb2-5" title="5">    bam_files, </a>
<a class="sourceLine" id="cb2-6" title="6">    <span class="dt">shift_ATAC_fragments =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb2-7" title="7">)</a>
<a class="sourceLine" id="cb2-8" title="8"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(bam_list) &lt;-<span class="st"> </span>bam_files</a></code></pre></div>
<p>For this vignette, relevant data is also hosted on the Ahringer server:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="co"># Import ATAC-seq fragments </span></a>
<a class="sourceLine" id="cb3-2" title="2">bam_list &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/connections.html">url</a></span>(<span class="st">'http://ahringerlab.com/VplotR/ATAC_PE_fragments.rds'</span>))</a>
<a class="sourceLine" id="cb3-3" title="3"><span class="kw"><a href="https://rdrr.io/r/base/lengths.html">lengths</a></span>(bam_list)</a></code></pre></div>
<p>We also need some relevant genomic locations. Let’s get different sets of <em>C. elegans</em> promoters.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="co"># Import annotated ce11 promoters </span></a>
<a class="sourceLine" id="cb4-2" title="2">proms_list &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/connections.html">url</a></span>(<span class="st">'http://ahringerlab.com/VplotR/promoters_Granges.rds'</span>))</a>
<a class="sourceLine" id="cb4-3" title="3"><span class="kw"><a href="https://rdrr.io/r/base/lengths.html">lengths</a></span>(proms_list)</a></code></pre></div>
</div>
<div id="check-distribution-of-fragment-sizes-over-a-specific-set-of-segments" class="section level2">
<h2 class="hasAnchor">
<a href="#check-distribution-of-fragment-sizes-over-a-specific-set-of-segments" class="anchor"></a>2. Check distribution of fragment sizes over a specific set of segments</h2>
<p>Let’s have a look at the distribution of fragment sizes in each ATAC-seq library, over the different classes of promoters we have. For this, we can use the <code><a href="../reference/getFragmentsDistribution.html">getFragmentsDistribution()</a></code> function.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="co"># For each bam file, plot its distribution of fragment sizes over the tissue-specific REs or ubiquitous REs</span></a>
<a class="sourceLine" id="cb5-2" title="2">sizes &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/parallel/mclapply.html">mclapply</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(bam_list), <span class="cf">function</span>(TISSUE) {</a>
<a class="sourceLine" id="cb5-3" title="3">    <span class="kw"><a href="../reference/getFragmentsDistribution.html">getFragmentsDistribution</a></span>(bam_list[[TISSUE]], proms_list, <span class="dt">roll =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb5-4" title="4">}, <span class="dt">mc.cores =</span> <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(bam_list)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-5" title="5"><span class="st">    </span><span class="kw"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(bam_list)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-6" title="6"><span class="st">    </span><span class="kw"><a href="../reference/namedListToLongFormat.html">namedListToLongFormat</a></span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-7" title="7"><span class="st">    </span><span class="kw"><a href="../reference/setColNames.html">setColNames</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'REs'</span>, <span class="st">'Fragment_size'</span>, <span class="st">'Number'</span>, <span class="st">'Promoters'</span>, <span class="st">'Sample'</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-8" title="8"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">Promoters =</span> <span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(Promoters, <span class="dt">levels =</span> <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(proms_list))) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-9" title="9"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">max_number =</span> <span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(Sample), <span class="cf">function</span>(TISSUE) {</a>
<a class="sourceLine" id="cb5-10" title="10">        <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(Number[Sample <span class="op">==</span><span class="st"> </span>TISSUE]), <span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(Sample <span class="op">==</span><span class="st"> </span>TISSUE))</a>
<a class="sourceLine" id="cb5-11" title="11">    }))) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-12" title="12"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(Sample <span class="op">!=</span><span class="st"> 'mixed'</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-13" title="13"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(Promoters <span class="op">==</span><span class="st"> 'Ubiq._proms'</span> <span class="op">|</span><span class="st"> </span>(<span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">'_proms'</span>, <span class="st">''</span>, .<span class="op">$</span>Promoters) <span class="op">==</span><span class="st"> </span>.<span class="op">$</span>Sample))</a>
<a class="sourceLine" id="cb5-14" title="14"><span class="co"># Plot results</span></a>
<a class="sourceLine" id="cb5-15" title="15"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(sizes, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> Fragment_size, <span class="dt">y =</span> Number, <span class="dt">color =</span> Promoters)) <span class="op">+</span></a>
<a class="sourceLine" id="cb5-16" title="16"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_blank.html">geom_blank</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">y =</span> max_number)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-17" title="17"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb5-18" title="18"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb5-19" title="19"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(</a>
<a class="sourceLine" id="cb5-20" title="20">        <span class="dt">title =</span> <span class="st">'Distribution of fragment sizes'</span>, </a>
<a class="sourceLine" id="cb5-21" title="21">        <span class="dt">x =</span> <span class="st">'Fragment size'</span>,</a>
<a class="sourceLine" id="cb5-22" title="22">        <span class="dt">y =</span> <span class="st">'# of fragments'</span>,</a>
<a class="sourceLine" id="cb5-23" title="23">        <span class="dt">color =</span> <span class="st">'Promoters classes'</span></a>
<a class="sourceLine" id="cb5-24" title="24">    ) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-25" title="25"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(Sample<span class="op">~</span>Promoters, <span class="dt">scales =</span> <span class="st">"free"</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-26" title="26"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(</a>
<a class="sourceLine" id="cb5-27" title="27">        <span class="dt">values =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'#991919'</span>, <span class="st">'#1232D9'</span>, <span class="st">'#3B9B46'</span>, <span class="st">'#D99B12'</span>, <span class="st">'#9e9e9e'</span>, <span class="st">'#D912D4'</span>)</a>
<a class="sourceLine" id="cb5-28" title="28">    )</a></code></pre></div>
<p>We can already notice that there is a lack of second/third/etc. nucleosome “bump” in somatic-specific ATAC-seq data over somatic REs. However, a more in-depth view of nucleosome binding would be desirable, to better understand the molecular context of promoter organization.</p>
</div>
<div id="vplots" class="section level2">
<h2 class="hasAnchor">
<a href="#vplots" class="anchor"></a>3. Vplots</h2>
<div id="quick-vplot" class="section level3">
<h3 class="hasAnchor">
<a href="#quick-vplot" class="anchor"></a>3.1 Quick Vplot</h3>
<p>Particularly, <em>where</em> fragments are mapping is an information as important as <em>how long</em> they are. The fragment length distribution plots generated hereabove do not show the former parameter (where each fragment is mapping). For this reason, we created a function to generate V-plots, to show distribution of ATAC fragments of variable sizes over a set of regions of interest. Such plot shows:</p>
<ul>
<li>On the x axis: the position of the ATAC fragments (relative to the center of the loci of interest);</li>
<li>On the y axis: the size of the ATAC fragments</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1">Vmat &lt;-<span class="st"> </span>bam_list[[<span class="st">'Neurons'</span>]] <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-2" title="2"><span class="st">    </span><span class="kw"><a href="../reference/computeVmat.html">computeVmat</a></span>(proms_list[[<span class="st">'Neurons_proms'</span>]]) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-3" title="3"><span class="st">    </span><span class="kw"><a href="../reference/normalizeVmat.html">normalizeVmat</a></span>(<span class="dt">normFun =</span> <span class="st">'pctsum'</span>, <span class="dt">roll =</span> <span class="dv">3</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-4" title="4"><span class="st">    </span><span class="kw"><a href="../reference/plotVmat.html">plotVmat</a></span>(</a>
<a class="sourceLine" id="cb6-5" title="5">        <span class="dt">main =</span> <span class="st">'Neurons-spe. ATAC fragments over Neurons-spe. proms'</span></a>
<a class="sourceLine" id="cb6-6" title="6">    )</a>
<a class="sourceLine" id="cb6-7" title="7"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span>(<span class="st">'examples/Vmat_Neurons-over-neurons-REs.pdf'</span>)</a>
<a class="sourceLine" id="cb6-8" title="8"><span class="co"># Or do all at once in one function:</span></a>
<a class="sourceLine" id="cb6-9" title="9">neurons_atac_neurons_proms &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plotVmat.html">plotVmat</a></span>(</a>
<a class="sourceLine" id="cb6-10" title="10">    bam_list[[<span class="st">'Neurons'</span>]], </a>
<a class="sourceLine" id="cb6-11" title="11">    proms_list[[<span class="st">'Neurons_proms'</span>]], </a>
<a class="sourceLine" id="cb6-12" title="12">    <span class="dt">estimate_background =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb6-13" title="13">    <span class="dt">main =</span> <span class="st">'Neurons-spe. ATAC fragments over Neurons-spe. proms'</span>, </a>
<a class="sourceLine" id="cb6-14" title="14">)</a>
<a class="sourceLine" id="cb6-15" title="15">gl_atac_gl_proms &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plotVmat.html">plotVmat</a></span>(</a>
<a class="sourceLine" id="cb6-16" title="16">    bam_list[[<span class="st">'Germline'</span>]], </a>
<a class="sourceLine" id="cb6-17" title="17">    proms_list[[<span class="st">'Germline_proms'</span>]], </a>
<a class="sourceLine" id="cb6-18" title="18">    <span class="dt">estimate_background =</span> <span class="ot">FALSE</span>, </a>
<a class="sourceLine" id="cb6-19" title="19">    <span class="dt">main =</span> <span class="st">'Germline-spe. ATAC fragments over Germline-spe. proms'</span>, </a>
<a class="sourceLine" id="cb6-20" title="20">)</a>
<a class="sourceLine" id="cb6-21" title="21">plots &lt;-<span class="st"> </span>cowplot<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(gl_atac_gl_proms, neurons_atac_neurons_proms)</a>
<a class="sourceLine" id="cb6-22" title="22"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span>(<span class="st">'examples/Germline_Neurons_Vmat-comparison.pdf'</span>, <span class="dt">height =</span> <span class="dv">6</span>, <span class="dt">width =</span> <span class="dv">12</span>, plots)</a></code></pre></div>
</div>
<div id="normalized-vplots" class="section level3">
<h3 class="hasAnchor">
<a href="#normalized-vplots" class="anchor"></a>3.2 Normalized Vplots</h3>
<p>Ideally, one would take into account the background ATAC-seq fragments to better estimate the molecular context in each condition. Samples might have an background fragment size higher than others and this could affect the appearance of un-normalized Vplots.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">Vmat_background &lt;-<span class="st"> </span><span class="kw"><a href="../reference/computeVmat.html">computeVmat</a></span>(</a>
<a class="sourceLine" id="cb7-2" title="2">    bam_list[[<span class="st">'Neurons'</span>]], </a>
<a class="sourceLine" id="cb7-3" title="3">    <span class="kw"><a href="../reference/shuffleGRanges.html">shuffleGRanges</a></span>(proms_list[[<span class="st">'Neurons_proms'</span>]])</a>
<a class="sourceLine" id="cb7-4" title="4">)</a>
<a class="sourceLine" id="cb7-5" title="5">Vmat &lt;-<span class="st"> </span>bam_list[[<span class="st">'Neurons'</span>]] <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb7-6" title="6"><span class="st">    </span><span class="kw"><a href="../reference/computeVmat.html">computeVmat</a></span>(proms_list[[<span class="st">'Neurons_proms'</span>]]) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb7-7" title="7"><span class="st">    </span><span class="kw"><a href="../reference/normalizeVmat.html">normalizeVmat</a></span>(</a>
<a class="sourceLine" id="cb7-8" title="8">        <span class="dt">background =</span> Vmat_background, </a>
<a class="sourceLine" id="cb7-9" title="9">        <span class="dt">scale =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb7-10" title="10">        <span class="dt">normFun =</span> <span class="st">'pctsum'</span>, </a>
<a class="sourceLine" id="cb7-11" title="11">        <span class="dt">roll =</span> <span class="dv">3</span></a>
<a class="sourceLine" id="cb7-12" title="12">    ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb7-13" title="13"><span class="st">    </span><span class="kw"><a href="../reference/plotVmat.html">plotVmat</a></span>(</a>
<a class="sourceLine" id="cb7-14" title="14">        <span class="dt">main =</span> <span class="st">'Neurons-spe. ATAC fragments over Neurons-spe. proms'</span></a>
<a class="sourceLine" id="cb7-15" title="15">    )</a>
<a class="sourceLine" id="cb7-16" title="16"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span>(<span class="st">'examples/Vmat_Neurons-over-neurons-REs_NORMALIZED.pdf'</span>)</a>
<a class="sourceLine" id="cb7-17" title="17"><span class="co"># An easier way to include such background normalization is to use the </span></a>
<a class="sourceLine" id="cb7-18" title="18"><span class="co"># argument `estimate_background = TRUE`</span></a>
<a class="sourceLine" id="cb7-19" title="19"><span class="co"># Scales can also be adjusted using the `breaks` argument.</span></a>
<a class="sourceLine" id="cb7-20" title="20">neurons_atac_neurons_proms_NORM &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plotVmat.html">plotVmat</a></span>(</a>
<a class="sourceLine" id="cb7-21" title="21">    bam_list[[<span class="st">'Neurons'</span>]], </a>
<a class="sourceLine" id="cb7-22" title="22">    proms_list[[<span class="st">'Neurons_proms'</span>]], </a>
<a class="sourceLine" id="cb7-23" title="23">    <span class="dt">estimate_background =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb7-24" title="24">    <span class="dt">breaks =</span> <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="dv">0</span>, <span class="dv">12</span>, <span class="dt">length.out =</span> <span class="dv">61</span>), </a>
<a class="sourceLine" id="cb7-25" title="25">    <span class="dt">main =</span> <span class="st">'Neurons-spe. ATAC fragments over Neurons-spe. proms (Norm.)'</span></a>
<a class="sourceLine" id="cb7-26" title="26">)</a>
<a class="sourceLine" id="cb7-27" title="27">gl_atac_gl_proms_NORM &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plotVmat.html">plotVmat</a></span>(</a>
<a class="sourceLine" id="cb7-28" title="28">    bam_list[[<span class="st">'Germline'</span>]], </a>
<a class="sourceLine" id="cb7-29" title="29">    proms_list[[<span class="st">'Germline_proms'</span>]], </a>
<a class="sourceLine" id="cb7-30" title="30">    <span class="dt">estimate_background =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb7-31" title="31">    <span class="dt">breaks =</span> <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="dv">0</span>, <span class="dv">12</span>, <span class="dt">length.out =</span> <span class="dv">61</span>), </a>
<a class="sourceLine" id="cb7-32" title="32">    <span class="dt">main =</span> <span class="st">'Germline-spe. ATAC fragments over Germline-spe. proms (Norm.)'</span></a>
<a class="sourceLine" id="cb7-33" title="33">)</a>
<a class="sourceLine" id="cb7-34" title="34">plots &lt;-<span class="st"> </span>cowplot<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(gl_atac_gl_proms_NORM, neurons_atac_neurons_proms_NORM)</a>
<a class="sourceLine" id="cb7-35" title="35"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span>(<span class="st">'examples/Germline_Neurons_NORMALIZED-Vmat-comparison.pdf'</span>, <span class="dt">height =</span> <span class="dv">6</span>, <span class="dt">width =</span> <span class="dv">12</span>, plots)</a></code></pre></div>
<p>To compute many different Vplots simultaneously, one can pass the two main arguments (bam_granges and granges) to the plotVmat function using a named list.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1">list_params &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(</a>
<a class="sourceLine" id="cb8-2" title="2">    <span class="st">"Germline ATAC-seq over Germline proms."</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(bam_list[[<span class="st">'Germline'</span>]], proms_list[[<span class="st">'Germline_proms'</span>]]),</a>
<a class="sourceLine" id="cb8-3" title="3">    <span class="st">"Germline ATAC-seq over Ubiq. proms."</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(bam_list[[<span class="st">'Germline'</span>]], proms_list[[<span class="st">'Ubiq._proms'</span>]]),</a>
<a class="sourceLine" id="cb8-4" title="4">    <span class="st">"Neurons ATAC-seq over Neurons proms."</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(bam_list[[<span class="st">'Neurons'</span>]], proms_list[[<span class="st">'Neurons_proms'</span>]]),</a>
<a class="sourceLine" id="cb8-5" title="5">    <span class="st">"Neurons ATAC-seq over Ubiq. proms."</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(bam_list[[<span class="st">'Neurons'</span>]], proms_list[[<span class="st">'Ubiq._proms'</span>]]),</a>
<a class="sourceLine" id="cb8-6" title="6">    <span class="st">"Muscle ATAC-seq over Muscle proms."</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(bam_list[[<span class="st">'Muscle'</span>]], proms_list[[<span class="st">'Muscle_proms'</span>]]),</a>
<a class="sourceLine" id="cb8-7" title="7">    <span class="st">"Muscle ATAC-seq over Ubiq. proms."</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(bam_list[[<span class="st">'Muscle'</span>]], proms_list[[<span class="st">'Ubiq._proms'</span>]]),</a>
<a class="sourceLine" id="cb8-8" title="8">    <span class="st">"Hypod. ATAC-seq over Hypod. proms."</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(bam_list[[<span class="st">'Hypod.'</span>]], proms_list[[<span class="st">'Hypod._proms'</span>]]),</a>
<a class="sourceLine" id="cb8-9" title="9">    <span class="st">"Hypod. ATAC-seq over Ubiq. proms."</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(bam_list[[<span class="st">'Hypod.'</span>]], proms_list[[<span class="st">'Ubiq._proms'</span>]]),</a>
<a class="sourceLine" id="cb8-10" title="10">    <span class="st">"Intest. ATAC-seq over Intest. proms."</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(bam_list[[<span class="st">'Intest.'</span>]], proms_list[[<span class="st">'Intest._proms'</span>]]),</a>
<a class="sourceLine" id="cb8-11" title="11">    <span class="st">"Intest. ATAC-seq over Ubiq. proms."</span> =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(bam_list[[<span class="st">'Intest.'</span>]], proms_list[[<span class="st">'Ubiq._proms'</span>]])</a>
<a class="sourceLine" id="cb8-12" title="12">)</a>
<a class="sourceLine" id="cb8-13" title="13">plots &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plotVmat.html">plotVmat</a></span>(</a>
<a class="sourceLine" id="cb8-14" title="14">    list_params, </a>
<a class="sourceLine" id="cb8-15" title="15">    <span class="dt">xlim =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="op">-</span><span class="dv">200</span>, <span class="dv">200</span>), </a>
<a class="sourceLine" id="cb8-16" title="16">    <span class="dt">estimate_background =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb8-17" title="17">    <span class="dt">cores =</span> <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(list_params)</a>
<a class="sourceLine" id="cb8-18" title="18">)</a>
<a class="sourceLine" id="cb8-19" title="19">p &lt;-<span class="st"> </span>plots <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-20" title="20"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span>Cond., <span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">dir =</span> <span class="st">'v'</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-21" title="21"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">legend.position =</span> <span class="st">'bottom'</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb8-22" title="22"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">panel.spacing =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/reexports.html">unit</a></span>(<span class="dv">1</span>, <span class="st">"lines"</span>))</a>
<a class="sourceLine" id="cb8-23" title="23"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span>(<span class="st">'examples/Comparison_tissue-specific-normalized-Vmats.pdf'</span>, <span class="dt">height =</span> <span class="fl">7.2</span>, <span class="dt">width =</span> <span class="dv">18</span>)</a></code></pre></div>
</div>
</div>
<div id="real-case-uses" class="section level2">
<h2 class="hasAnchor">
<a href="#real-case-uses" class="anchor"></a>4 Real case uses</h2>
<div id="real-case-use-1-compare-nucleosome-patterns-over-tissue-specific-promoters-or-enhancers" class="section level3">
<h3 class="hasAnchor">
<a href="#real-case-use-1-compare-nucleosome-patterns-over-tissue-specific-promoters-or-enhancers" class="anchor"></a>4.1. Real case use #1: Compare nucleosome patterns over tissue-specific promoters or enhancers</h3>
<p>Let’s have a look at tissue-specific classes of promoters and the associated Vplots, in a “programmatic” approach. We already presented the possibility to pass a list of arguments to plotVmat (see section 2.2). This allows a person to plot multiple Vplots at once. Another way is to rely on parallelized computations of individual Vmats. Such simpler approach generating a list of plots allows the user to personalize each plot individually.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1">plotlist &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/parallel/mclapply.html">mclapply</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span>(proms_list), <span class="cf">function</span>(K) {</a>
<a class="sourceLine" id="cb9-2" title="2">    plot.proms &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/parallel/mclapply.html">mclapply</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span>(bam_list), <span class="cf">function</span>(B) {</a>
<a class="sourceLine" id="cb9-3" title="3">        <span class="kw"><a href="https://rdrr.io/r/base/message.html">message</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(bam_list)[B])</a>
<a class="sourceLine" id="cb9-4" title="4">        <span class="kw"><a href="../reference/plotVmat.html">plotVmat</a></span>(</a>
<a class="sourceLine" id="cb9-5" title="5">            bam_list[[B]], </a>
<a class="sourceLine" id="cb9-6" title="6">            proms_list[[K]],</a>
<a class="sourceLine" id="cb9-7" title="7">            <span class="dt">estimate_background =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb9-8" title="8">            <span class="dt">main =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(bam_list)[B], <span class="st">'-spe. ATAC-seq over '</span>, <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(proms_list)[K], <span class="st">' proms'</span>), </a>
<a class="sourceLine" id="cb9-9" title="9">            <span class="dt">breaks =</span> <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="dv">0</span>, <span class="dv">12</span>, <span class="dt">length.out =</span> <span class="dv">61</span>), </a>
<a class="sourceLine" id="cb9-10" title="10">            <span class="dt">ylim =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">55</span>, <span class="dv">300</span>)</a>
<a class="sourceLine" id="cb9-11" title="11">        )</a>
<a class="sourceLine" id="cb9-12" title="12">    }, <span class="dt">mc.cores =</span> <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(bam_list))</a>
<a class="sourceLine" id="cb9-13" title="13">    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(plot.proms)</a>
<a class="sourceLine" id="cb9-14" title="14">}, <span class="dt">mc.cores =</span> <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(proms_list))</a>
<a class="sourceLine" id="cb9-15" title="15">plots &lt;-<span class="st"> </span>purrr<span class="op">::</span><span class="kw"><a href="https://purrr.tidyverse.org/reference/flatten.html">flatten</a></span>(plotlist)</a>
<a class="sourceLine" id="cb9-16" title="16">p &lt;-<span class="st"> </span>ggpubr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/ggpubr/man/ggarrange.html">ggarrange</a></span>(</a>
<a class="sourceLine" id="cb9-17" title="17">    <span class="dt">plotlist =</span> plots, </a>
<a class="sourceLine" id="cb9-18" title="18">    <span class="dt">nrow =</span> <span class="dv">6</span>, <span class="dt">ncol =</span> <span class="dv">6</span>, </a>
<a class="sourceLine" id="cb9-19" title="19">    <span class="dt">common.legend =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb9-20" title="20">    <span class="dt">legend =</span> <span class="st">"bottom"</span></a>
<a class="sourceLine" id="cb9-21" title="21">)</a>
<a class="sourceLine" id="cb9-22" title="22"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span>(<span class="st">'examples/Comparison-tissue-specific-normalized-Vmats_all-combinations.pdf'</span>, <span class="dt">height =</span> <span class="dv">45</span>, <span class="dt">width =</span> <span class="dv">40</span>)</a></code></pre></div>
<p>It is striking that Germline promoters are characterized by prominent flanking nucleosomes, while somatic promoters are not. Ubiquitous promoters, on the other hand, are flanked by nucleosomes in all tissues.</p>
</div>
<div id="real-case-use-2-nucleosome-enrichment-estimation" class="section level3">
<h3 class="hasAnchor">
<a href="#real-case-use-2-nucleosome-enrichment-estimation" class="anchor"></a>4.2. Real case use #2: Nucleosome enrichment estimation</h3>
<p>We can estimate nucleosome enrichment using the <code><a href="../reference/nucleosomeEnrichment.html">nucleosomeEnrichment()</a></code> function. This function is used to quantify the local enrichment of a nucleosome at a certain distance from the midpoint of the genomic ranges of interest.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">list_scores &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/parallel/mclapply.html">mclapply</a></span>(</a>
<a class="sourceLine" id="cb10-2" title="2">    <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Germline"</span>, <span class="st">"Neurons"</span>, <span class="st">"Muscle"</span>, <span class="st">"Hypod."</span>, <span class="st">"Intest."</span>), </a>
<a class="sourceLine" id="cb10-3" title="3">    <span class="cf">function</span>(TISSUE) {</a>
<a class="sourceLine" id="cb10-4" title="4">        <span class="kw"><a href="https://rdrr.io/r/base/message.html">message</a></span>(<span class="st">'&gt;&gt; '</span>, TISSUE)</a>
<a class="sourceLine" id="cb10-5" title="5">        nucenrich_tissue_spe_proms &lt;-<span class="st"> </span><span class="kw"><a href="../reference/nucleosomeEnrichment.html">nucleosomeEnrichment</a></span>(</a>
<a class="sourceLine" id="cb10-6" title="6">            <span class="dt">bam_granges =</span> bam_list[[TISSUE]], </a>
<a class="sourceLine" id="cb10-7" title="7">            <span class="dt">granges =</span> proms_list[[<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(TISSUE, <span class="st">'_proms'</span>)]], </a>
<a class="sourceLine" id="cb10-8" title="8">            <span class="dt">estimate_background =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb10-9" title="9">            <span class="dt">verbose =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb10-10" title="10">        )</a>
<a class="sourceLine" id="cb10-11" title="11">        nucenrich_ubiq_proms &lt;-<span class="st"> </span><span class="kw"><a href="../reference/nucleosomeEnrichment.html">nucleosomeEnrichment</a></span>(</a>
<a class="sourceLine" id="cb10-12" title="12">            <span class="dt">bam_granges =</span> bam_list[[TISSUE]], </a>
<a class="sourceLine" id="cb10-13" title="13">            <span class="dt">granges =</span> proms_list[[<span class="st">'Ubiq._proms'</span>]], </a>
<a class="sourceLine" id="cb10-14" title="14">            <span class="dt">estimate_background =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb10-15" title="15">            <span class="dt">verbose =</span> <span class="ot">TRUE</span></a>
<a class="sourceLine" id="cb10-16" title="16">        )</a>
<a class="sourceLine" id="cb10-17" title="17">        <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(</a>
<a class="sourceLine" id="cb10-18" title="18">            <span class="st">'tissue-spe-proms'</span> =<span class="st"> </span>nucenrich_tissue_spe_proms, </a>
<a class="sourceLine" id="cb10-19" title="19">            <span class="st">'ubiq-proms'</span> =<span class="st"> </span>nucenrich_ubiq_proms</a>
<a class="sourceLine" id="cb10-20" title="20">        ))</a>
<a class="sourceLine" id="cb10-21" title="21">    }, </a>
<a class="sourceLine" id="cb10-22" title="22">    <span class="dt">mc.cores =</span> <span class="dv">5</span></a>
<a class="sourceLine" id="cb10-23" title="23">) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Germline"</span>, <span class="st">"Neurons"</span>, <span class="st">"Muscle"</span>, <span class="st">"Hypod."</span>, <span class="st">"Intest."</span>))</a>
<a class="sourceLine" id="cb10-24" title="24">nucenrich_scores &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(</a>
<a class="sourceLine" id="cb10-25" title="25">    <span class="dt">tissue =</span> <span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(list_scores), <span class="dt">each =</span> <span class="dv">2</span>), <span class="dt">levels =</span> <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(list_scores)), </a>
<a class="sourceLine" id="cb10-26" title="26">    <span class="dt">promoters =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(list_scores), <span class="st">' promoters'</span>), <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">'Ubiq. promoters'</span>, <span class="dv">5</span>))), </a>
<a class="sourceLine" id="cb10-27" title="27">    <span class="dt">promoters2 =</span> <span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.5</span>, <span class="fl">0.8</span>), <span class="dv">5</span>)),</a>
<a class="sourceLine" id="cb10-28" title="28">    <span class="dt">score =</span> <span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(list_scores, <span class="cf">function</span>(Vmat) <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(Vmat[[<span class="dv">1</span>]]<span class="op">$</span>fisher_test<span class="op">$</span>estimate, Vmat[[<span class="dv">2</span>]]<span class="op">$</span>fisher_test<span class="op">$</span>estimate))), </a>
<a class="sourceLine" id="cb10-29" title="29">    <span class="dt">is_ubiq =</span> <span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Over tissue-specific promoters'</span>, <span class="st">'Over ubiquitous promoters'</span>), <span class="dv">5</span>))</a>
<a class="sourceLine" id="cb10-30" title="30">)</a>
<a class="sourceLine" id="cb10-31" title="31">p &lt;-<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(nucenrich_scores, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(</a>
<a class="sourceLine" id="cb10-32" title="32">    <span class="dt">x =</span> tissue, </a>
<a class="sourceLine" id="cb10-33" title="33">    <span class="dt">y =</span> score, </a>
<a class="sourceLine" id="cb10-34" title="34">    <span class="dt">fill =</span> tissue, </a>
<a class="sourceLine" id="cb10-35" title="35">    <span class="dt">group =</span> is_ubiq</a>
<a class="sourceLine" id="cb10-36" title="36">)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb10-37" title="37"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb10-38" title="38"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span>is_ubiq, <span class="dt">nrow =</span> <span class="dv">2</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb10-39" title="39"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb10-40" title="40"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span>(<span class="dt">title =</span> <span class="st">'Flanking nucleosome enrichment score'</span>, <span class="dt">y =</span> <span class="st">'Enrichment score'</span>, <span class="dt">x =</span> <span class="st">'Tissue-specific data'</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb10-41" title="41"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb10-42" title="42"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">legend.position =</span> <span class="st">'none'</span>)</a>
<a class="sourceLine" id="cb10-43" title="43"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span>(<span class="st">'examples/Comparison_tissue-specific-nucleosome-enrichment.pdf'</span>, <span class="dt">height =</span> <span class="dv">5</span>, <span class="dt">width =</span> <span class="dv">5</span>)</a></code></pre></div>
<p>This confirms what the Vplots suggest: flanking nucleosomes are well enriched at ubiquitous promoters in all tissues. Among tissue-specific promoters, only germline promoters have a significant enrichment of flanking nucleosome.</p>
</div>
<div id="real-case-use-3-use-mnase-seq-data" class="section level3">
<h3 class="hasAnchor">
<a href="#real-case-use-3-use-mnase-seq-data" class="anchor"></a>4 .3. Real case use #3: Use MNase-seq data</h3>
<p>We can also use MNase data (only this data comes from mixed tissues) to look at nucleosome patterns.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1"><span class="co"># Read the bam files from local folder: </span></a>
<a class="sourceLine" id="cb11-2" title="2"><span class="co">### mnase_files &lt;- list.files(</span></a>
<a class="sourceLine" id="cb11-3" title="3"><span class="co">###     '~/20181127_PE-mnase-mapping/_bam-files', </span></a>
<a class="sourceLine" id="cb11-4" title="4"><span class="co">###     pattern = 'rep-1.*q10.bam', </span></a>
<a class="sourceLine" id="cb11-5" title="5"><span class="co">###     full.names = TRUE</span></a>
<a class="sourceLine" id="cb11-6" title="6"><span class="co">### )</span></a>
<a class="sourceLine" id="cb11-7" title="7"><span class="co">### MNase_mixed &lt;- importPEBamFiles(</span></a>
<a class="sourceLine" id="cb11-8" title="8"><span class="co">###     mnase_files, </span></a>
<a class="sourceLine" id="cb11-9" title="9"><span class="co">###     where = GenomicRanges::reduce(GenomicRanges::resize(ce_proms, width = 2000, fix = 'center')), </span></a>
<a class="sourceLine" id="cb11-10" title="10"><span class="co">###     shift_ATAC_fragments = FALSE</span></a>
<a class="sourceLine" id="cb11-11" title="11"><span class="co">### ) %&gt;% </span></a>
<a class="sourceLine" id="cb11-12" title="12"><span class="co">###     setNames(mnase_files %&gt;% basename() %&gt;% gsub('.map_pe\\^rm_chrM\\^rm_blacklist\\^q10.bam', '', .)) %&gt;% </span></a>
<a class="sourceLine" id="cb11-13" title="13"><span class="co">###     '['(names(.) %&gt;% gsub('YA_mnase_rep-1_|U', '', .) %&gt;% as.numeric() %&gt;% order())</span></a>
<a class="sourceLine" id="cb11-14" title="14"><span class="co"># Or simply: </span></a>
<a class="sourceLine" id="cb11-15" title="15">MNase_mixed &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span>(</a>
<a class="sourceLine" id="cb11-16" title="16">    <span class="kw"><a href="https://rdrr.io/r/base/connections.html">url</a></span>(</a>
<a class="sourceLine" id="cb11-17" title="17">        <span class="st">'http://ahringerlab.com/VplotR/MNase_PE_fragments.rds'</span></a>
<a class="sourceLine" id="cb11-18" title="18">    )</a>
<a class="sourceLine" id="cb11-19" title="19">)</a>
<a class="sourceLine" id="cb11-20" title="20"><span class="co"># Then plot Vmats</span></a>
<a class="sourceLine" id="cb11-21" title="21">list_params &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span>(MNase_mixed), <span class="cf">function</span>(B) {</a>
<a class="sourceLine" id="cb11-22" title="22">    <span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span>(proms_list), <span class="cf">function</span>(K) {</a>
<a class="sourceLine" id="cb11-23" title="23">        <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(</a>
<a class="sourceLine" id="cb11-24" title="24">            MNase_mixed[[B]], </a>
<a class="sourceLine" id="cb11-25" title="25">            <span class="kw">resize</span>(proms_list[[K]], <span class="dv">2000</span>, <span class="dt">fix =</span> <span class="st">'center'</span>)</a>
<a class="sourceLine" id="cb11-26" title="26">        )</a>
<a class="sourceLine" id="cb11-27" title="27">    }) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/setNames.html">setNames</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">'MNase ('</span>, <span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">'YA_mnase_rep-1_'</span>, <span class="st">''</span>, <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(MNase_mixed))[B], <span class="st">') coverage </span><span class="ch">\n</span><span class="st">over '</span>, <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(proms_list), <span class="st">' TSSs'</span>))</a>
<a class="sourceLine" id="cb11-28" title="28">}) <span class="op">%&gt;%</span><span class="st"> </span>purrr<span class="op">::</span><span class="kw"><a href="https://purrr.tidyverse.org/reference/flatten.html">flatten</a></span>()</a>
<a class="sourceLine" id="cb11-29" title="29">plots &lt;-<span class="st"> </span><span class="kw"><a href="../reference/plotVmat.html">plotVmat</a></span>(</a>
<a class="sourceLine" id="cb11-30" title="30">    list_params, </a>
<a class="sourceLine" id="cb11-31" title="31">    <span class="dt">XAXIS.CUTOFF =</span> <span class="dv">2000</span>, </a>
<a class="sourceLine" id="cb11-32" title="32">    <span class="dt">xlim =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="op">-</span><span class="dv">1000</span>, <span class="dv">1000</span>), </a>
<a class="sourceLine" id="cb11-33" title="33">    <span class="dt">ylim =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">75</span>, <span class="dv">200</span>),</a>
<a class="sourceLine" id="cb11-34" title="34">    <span class="dt">breaks =</span> <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="dv">0</span>, <span class="dv">15</span>, <span class="dt">length.out =</span> <span class="dv">60</span>), </a>
<a class="sourceLine" id="cb11-35" title="35">    <span class="dt">cores =</span> <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(list_params)</a>
<a class="sourceLine" id="cb11-36" title="36">)</a>
<a class="sourceLine" id="cb11-37" title="37">p &lt;-<span class="st"> </span>plots <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb11-38" title="38"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(<span class="op">~</span>Cond., <span class="dt">ncol =</span> <span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(proms_list), <span class="dt">dir =</span> <span class="st">'h'</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb11-39" title="39"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">legend.position =</span> <span class="st">'bottom'</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb11-40" title="40"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">panel.spacing =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/reexports.html">unit</a></span>(<span class="dv">1</span>, <span class="st">"lines"</span>)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb11-41" title="41"><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html">coord_fixed</a></span>(<span class="dt">ratio =</span> <span class="dv">7</span>)</a>
<a class="sourceLine" id="cb11-42" title="42"><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html">ggsave</a></span>(<span class="st">'examples/MNase-coverage_proms.pdf'</span>, <span class="dt">height =</span> <span class="dv">15</span>, <span class="dt">width =</span> <span class="dv">18</span>)</a></code></pre></div>
<p>As expected, nucleosomal reads are mapped over flanking nucleosomes, specifically around ubiquitous and germline promoters. Arrays of nucleosomes are readily detected at these promoters.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#set-up-the-session">0. Set up the session</a></li>
      <li><a href="#import-reads-bam-files-and-genomic-loci-files">1. Import reads bam files and genomic loci files</a></li>
      <li><a href="#check-distribution-of-fragment-sizes-over-a-specific-set-of-segments">2. Check distribution of fragment sizes over a specific set of segments</a></li>
      <li>
<a href="#vplots">3. Vplots</a><ul class="nav nav-pills nav-stacked">
<li><a href="#quick-vplot">3.1 Quick Vplot</a></li>
      <li><a href="#normalized-vplots">3.2 Normalized Vplots</a></li>
      </ul>
</li>
      <li>
<a href="#real-case-uses">4 Real case uses</a><ul class="nav nav-pills nav-stacked">
<li><a href="#real-case-use-1-compare-nucleosome-patterns-over-tissue-specific-promoters-or-enhancers">4.1. Real case use #1: Compare nucleosome patterns over tissue-specific promoters or enhancers</a></li>
      <li><a href="#real-case-use-2-nucleosome-enrichment-estimation">4.2. Real case use #2: Nucleosome enrichment estimation</a></li>
      <li><a href="#real-case-use-3-use-mnase-seq-data">4 .3. Real case use #3: Use MNase-seq data</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://https://js2264.github.io/">Jacques Serizay</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
